File: flow-client/src/test/java/com/vaadin/client/flow/reactive/CountingComputation.java
/*
 * Copyright (C) 2000-2024 Vaadin Ltd
 *
 * This program is available under Vaadin Commercial License and Service Terms.
 *
 * See <https://vaadin.com/commercial-license-and-service-terms> for the full
 * license.
 */
package com.vaadin.client.flow.reactive;

import com.vaadin.client.Command;

public class CountingComputation extends Computation {

    private Command reader;

    public CountingComputation(ReactiveEventRouter<?, ?> eventRouter) {
        this(eventRouter::registerRead);
    }

    public CountingComputation(Command reader) {
        this.reader = reader;
    }

    private int count = 0;

    @Override
    protected void doRecompute() {
        count++;
        reader.execute();
    }

    public int getCount() {
        return count;
    }
}

File: flow-client/src/test/frontend/internal/client/flow/reactive/CountingComputation.ts
import { Computation } from '../../../../../../main/frontend/internal/client/flow/reactive/Computation';

// Mirrors the Java CountingComputation helper: counts recomputations and runs a
// reader on each.
export function countingComputation(reader: () => void): { computation: Computation; getCount: () => number } {
  let count = 0;
  const computation = new Computation(() => {
    count++;
    reader();
  });
  return { computation, getCount: () => count };
}
